test/preconditioner/batch_jacobi_kernels.cpp

```diff
@@ -0,0 +1,252 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/preconditioner/batch_jacobi_kernels.hpp"
+
+
+#include <limits>
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/batch_csr.hpp>
+#include <ginkgo/core/preconditioner/batch_jacobi.hpp>
+#include <ginkgo/core/solver/batch_bicgstab.hpp>
+
+
+#include "core/solver/batch_bicgstab_kernels.hpp"
+#include "core/test/utils.hpp"
+#include "core/test/utils/batch_helpers.hpp"
+#include "test/utils/executor.hpp"
+
+
+namespace detail {
+
+
+template <typename ValueType>
+void is_equivalent_to_ref(
+    std::unique_ptr<gko::batch::preconditioner::Jacobi<ValueType>> ref_prec,
+    std::unique_ptr<gko::batch::preconditioner::Jacobi<ValueType>> d_prec)
+{
+    auto ref = ref_prec->get_executor();
+    auto exec = d_prec->get_executor();
+    const auto nbatch = ref_prec->get_num_batch_items();
+    const auto num_blocks = ref_prec->get_num_blocks();
+    const auto block_pointers_ref = ref_prec->get_const_block_pointers();
+
+    const auto tol = 10 * r<ValueType>::value;
+
+    gko::array<int> d_block_pointers_copied_to_ref(ref, num_blocks + 1);
+    ref->copy_from(exec.get(), num_blocks + 1,
+                   d_prec->get_const_block_pointers(),
+                   d_block_pointers_copied_to_ref.get_data());
+
+    gko::array<int> d_block_cumul_storage_copied_to_ref(ref, num_blocks + 1);
+    ref->copy_from(exec.get(), num_blocks + 1,
+                   d_prec->get_const_blocks_cumulative_storage(),
+                   d_block_cumul_storage_copied_to_ref.get_data());
+
+    for (int batch_id = 0; batch_id < nbatch; batch_id++) {
+        for (int block_id = 0; block_id < num_blocks; block_id++) {
+            const auto bsize =
+                block_pointers_ref[block_id + 1] - block_pointers_ref[block_id];
+
+            const auto ref_dense_block_ptr =
+                ref_prec->get_const_blocks() +
+                gko::detail::batch_jacobi::get_global_block_offset(
+                    batch_id, ref_prec->get_num_blocks(), block_id,
+                    ref_prec->get_const_blocks_cumulative_storage());
+            const auto ref_stride = gko::detail::batch_jacobi::get_stride(
+                block_id, block_pointers_ref);
+            const auto d_dense_block_ptr =
+                d_prec->get_const_blocks() +
+                gko::detail::batch_jacobi::get_global_block_offset(
+                    batch_id, d_prec->get_num_blocks(), block_id,
+                    d_block_cumul_storage_copied_to_ref.get_const_data());
+            const auto d_stride = gko::detail::batch_jacobi::get_stride(
+                block_id, d_block_pointers_copied_to_ref.get_const_data());
+
+            for (int r = 0; r < bsize; r++) {
```

Comment:
maybe instead of comparing the values manually, create dense matrix views on the diagonal blocks and compare them with the normal macro.